### PR TITLE
Added Field Resourse (Feature/#109)

### LINKED ISSUE #61: Implement Field Resource
- fields / {id}
- fields /

### PR DESCRIPTION
**Pull request by AndriiKret**
- Added field route
- Added functions to create fields with option specific to types
- Added MA Schemes to fields with specific types
- Added NotSend Exceptions
- Added validations to incoming data

### CODE REVIEW

File: src/app/routers/field.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+"""
+from flask import request, jsonify
+from flask_restx import fields, Resource
+from flask_login import current_user
+from app import API
+from app.services.field_crud import FieldOperation, FieldService
+
+NS_FIELDS = API.namespace('fields', description='NgFg APIs')
+
+MAIN_MODEL = API.model('fields', {
```
Comment (yurdosii) [inline src/app/routers/field.py]: Rename to FIELD_MODEL ?

File: src/app/routers/field.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    def post(self):
+        """
+        Field POST method
+
+        :return: json
+        """
+        req = request.json
+        data = FieldOperation.create(**req)
+
+        return {'data': data, 'SUCCESS': 'OK'}
```
Comment (yurdosii) [inline src/app/routers/field.py]: Use jsonify(), it sets header 'application/json'

File: src/app/services/field_crud.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
+"""
+Field operations
+"""
+from app.services import FieldService, FieldRangeService, RangeService, \
+    ChoiceOptionService, SettingAutocompleteService
```
Comment (yurdosii) [inline src/app/services/field_crud.py]: Use tuple

File: src/app/services/range.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
 from app.helper.decorators import transaction_decorator
 from app.helper.errors import RangeNotExist
 from app.models import Range
+from app.models.range import RangeSchema
```
Comment (yurdosii) [inline src/app/services/range.py]: You can import RangeSchema from app/models/__init__.py

File: src/app/routers/field.py (hunk 2 above)
Comment (Howrran) [reply src/app/routers/field.py]: fixed

File: src/app/services/field_crud.py (hunk 3 above)
Comment (Howrran) [reply src/app/services/field_crud.py]: fixed

File: src/app/services/range.py (hunk 4 above)
Comment (Howrran) [reply src/app/services/range.py]: fixed

**Review by zzeropilot — DISMISSED** [review]

File: src/app/__init__.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 from flask import Flask, Blueprint
 from flask_restx import Api
 from flask_marshmallow import Marshmallow
+from oauthlib.oauth2 import WebApplicationClient
```
Comment (IvanKostiv) [inline src/app/__init__.py]: Old auth, delete this

File: src/app/services/field_crud.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
+"""
+Field operations
+"""
```
Comment (IvanKostiv) [inline src/app/services/field_crud.py]: Use FieldService

File: src/app/routers/field.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+})
+
+EXTENDED_MODEL = API.inherit('Extended_field', FIELD_MODEL, {
+    "range": fields.Nested(RANGE_MODEL),
+    "choice_options": fields.List(fields.String),
+    "setting_autocomplete": fields.Nested(AUTOCOMPLETE_MODEL)
+})
+
+
+@FIELDS_NS.route("")
```
Comment (yurdosii) [inline src/app/routers/field.py]: route("/")

File: src/app/routers/field.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    def post(self):
+        """
+        Field POST method
+
+        :return: json
+        """
+        req = request.json
+        data = FieldOperation.create(**req)
+
+        return jsonify(data)
```
Comment (yurdosii) [inline src/app/routers/field.py]: Validation?
- if create() returns None
- if request body is incorrect
- onwer_id in request body cannot differ from request user id

Comment (AndriiKret) [reply src/app/routers/field.py]: fixed

File: src/app/routers/field.py (hunk 1 above)
Comment (Howrran) [reply src/app/routers/field.py]: fixed

File: src/app/__init__.py (hunk 5 above)
Comment (Howrran) [reply src/app/__init__.py]: fixed

File: src/app/routers/field.py (hunk 7 above)
Comment (Howrran) [reply src/app/routers/field.py]: fixed

File: src/app/services/field_crud.py (hunk 6 above)
Comment (Howrran) [reply src/app/services/field_crud.py]: fixed

File: src/app/services/field.py [hunk 9]
```diff
@@ -113,3 +124,271 @@ def delete(field_id):
...
+        """
+
+        field = FieldService.create(name=name,
+                                    # pylint: disable=too-many-arguments
+                                    owner_id=owner_id,
+                                    field_type=field_type,
+                                    is_strict=is_strict)
+
+        print('we are here1')
+        print(field)
```
Comment (Howrran) [inline src/app/services/field.py]: ех, щас би прінти не забрати...

File: src/app/models/setting_autocomplete.py [hunk 10]
```diff
@@ -20,10 +21,29 @@ class SettingAutocomplete(AbstractModel):
...
+    Setting autocomplete schema
+    """
+
+    class Meta:
+        """
+        Setting autocomplete schema meta
+        """
+        fields = ("data_url", "sheet", "from_row", "to_row")
+
+    data_url = fields.Str(required=True)
```
Comment (zzeropilot) [inline src/app/models/setting_autocomplete.py]: why not fields.Url?

**Review by IvanKostiv — APPROVED** [review]

**Review by zzeropilot — APPROVED** [review]

**Review by Howrran — APPROVED** [review]

**Pull request merged**